File: Tools/VikingExtractor/Source/Miscellaneous.cpp
/*
    VikingExtractor, to recover images from Viking Lander operations.
    Copyright (C) 2010, 2011, 2012 Kshatra Corp <kip@thevertigo.com>.
    
    Public discussion on IRC available at #avaneya (irc.freenode.net) 
    or on the mailing list <avaneya@lists.avaneya.com>.

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

// Includes...
#include "Miscellaneous.h"
#include <cstdio>
#include <cmath>
#include <cassert>
#include <fstream>
#include <sstream>
#include <string>
#include <cstring>
#include <sys/stat.h>

// Using the standard name space...
using namespace std;

// Create a directory and all of its parents, if necessary...
bool CreateDirectoryRecursively(const std::string &Path)
{
    // Make a copy of the user's string, since we need to change it...
    string PathCopy = Path;

    // Remove trailing path separator...
    if(PathCopy.at(PathCopy.length() - 1) == '/')
        PathCopy.erase(PathCopy.length() - 1);

    // Create the directories from the parent to child order...
    for(string::const_iterator Iterator = PathCopy.begin(); Iterator != PathCopy.end(); ++Iterator)
    {
        // Found the next partial path, create...
        if(*Iterator == '/')
        {
            // Get path up to this point...
            const string PartialPath(PathCopy, 0, Iterator - PathCopy.begin());

            // If it doesn't already exist, create it...
            if(access(PartialPath.c_str(), F_OK) != 0)
                mkdir(PartialPath.c_str(), S_IRWXU);
        }
    }

    // Check if final nested directory exists, if not, create...
    if(access(PathCopy.c_str(), F_OK) != 0)
        return (mkdir(PathCopy.c_str(), S_IRWXU) == 0);
    
    // Otherwise path already exists...
    else
        return true;
}

// Convert a given Martian solar day in the range [1 .. n] to Ls angle...
float SolarDayToLs(const size_t SolarDay)
{
    // Bounds check...
    assert(SolarDay >= 1);

    // Variables...
    float Ls                        = 0.0f;
    float EccentricAnomaly          = 0.0f;
    float EccentricAnomalyDelta     = 0.0f;

    // Constants...
    const float SolsInYear          = 668.6f;
    const float PerihelionDay       = 485.35f;
    const float PerihelionLs        = 250.99f;
    const float Ecentricity         = 0.09340f;
    const float RadiansToDegrees    = 180.0f / M_PI;
    const float TimePerihelion      = 2 * M_PI * (1 - PerihelionLs / 360.0f);

    // Calculate mean anomaly...
    const float zz = (SolarDay - PerihelionDay) / SolsInYear;
    const float SignedMeanAnomaly = 2.0f * M_PI * (zz - round(zz));
    const float MeanAnomaly = abs(SignedMeanAnomaly);

    // Solve Kepler equation MeanAnomaly = EccentricAnomaly - ε * sin(EccentricAnomaly)
    // using Newton iterations...
    EccentricAnomaly = MeanAnomaly + Ecentricity * sin(MeanAnomaly);
    do
    {
        EccentricAnomalyDelta = -(EccentricAnomaly - Ecentricity * sin(EccentricAnomaly) - MeanAnomaly) / 
               (1.0f - Ecentricity * cos(EccentricAnomaly));
        EccentricAnomaly = EccentricAnomaly + EccentricAnomalyDelta;
    }
    while(EccentricAnomalyDelta > 1.0e-6f);
    
    if(SignedMeanAnomaly < 0.0f)
        EccentricAnomaly = -EccentricAnomaly;

    // Compute true anomaly, now that eccentric anomaly is known...
    const float TrueAnomaly = 
        2.0f * atan(sqrt((1.0f + Ecentricity) / 
        (1.0f - Ecentricity)) * tan(EccentricAnomaly / 2.0f));

    // Compute Ls...
    Ls = TrueAnomaly - TimePerihelion;
    if(Ls < 0.0f)
        Ls += 2.0f * M_PI;
    if(Ls > 2.0f * M_PI)
        Ls -= 2.0f * M_PI;

    // Convert Ls from radians into degrees and return...
    return (RadiansToDegrees * Ls);
}

File: Extras/Viking Lander Remastered/Launcher/Source/Miscellaneous.py
# VikingExtractor, to recover images from Viking Lander operations.
# Copyright (C) 2010-2013 Cartesian Theatre <info@cartesiantheatre.com>.
#
# Public discussion on IRC available at #avaneya (irc.freenode.net) or
# on the mailing list <avaneya@lists.avaneya.com>.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or 
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, but 
# WITHOUT ANY WARRANTY; without even the implied warranty of 
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU 
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.

"""Miscellaneous: Random routines that didn't belong anywhere else"""

# System imports...
from gi.repository import Gdk
from gi.repository import Gtk
from gi.repository import Gio
import os
import subprocess
import platform
import sys

# i18n...
import gettext
_ = gettext.gettext

# Privates...
_applicationName    = "Avaneya: Viking Lander Remastered DVD"
_versionMajor       = 0
_versionMinor       = 2
_versionShortString = "{0}.{1}".format(_versionMajor, _versionMinor)
_versionLongString  = "{0}, v{1}".format(_applicationName, _versionShortString)

# Get the machine architecture...
def getMachineArchitecture():

    # Retrieve...
    machineArchitecture = platform.machine()
    assert(machineArchitecture)
    
    # Adapt to Debian's sane nomenclature...
    if machineArchitecture == "x86_64":
        machineArchitecture = "amd64"
    elif machineArchitecture == "x86":
        machineArchitecture = "i386"

    # Return the machine architecture to the caller...
    return machineArchitecture

# Get the size of the monitor the caller's window is mostly visible on, or the
#  default monitor of the default screen of the root window if none given...
def getMonitorWithCursorSize():

    # Get the cursor position...
    display = Gdk.Display.get_default()
    (screen, cursorX, cursorY, rootWindow) = display.get_pointer()

    # Which monitor the user's cursor?
    monitorId = screen.get_monitor_at_point(cursorX, cursorY)

    # Get the geometry of that monitor...
    monitorRectangle = screen.get_monitor_geometry(monitorId)
    
    # Return the monitor's width and height...
    return (monitorRectangle.width, monitorRectangle.height)

# Get the operating system name...
def getOperatingSystem():

    # Retrieve...
    operatingSystem = platform.system()
    assert(operatingSystem)
    
    # Correct common misnomer. Kernel mistaken for entire OS...
    #  <https://www.gnu.org/gnu/gnu-linux-faq.html>
    if operatingSystem == "Linux":
        operatingSystem = "GNU"

    # Return operating system name to caller...
    return operatingSystem

# Get the application name and version...
def getApplicationName(): return _applicationName
def getLongVersionString(): return _versionLongString
def getShortVersionString(): return _versionShortString

# Check network connectivity to the internet and alert user if caller requests 
#  if no connection available...
def hasInternetConnection(alertUser, parent, unqueriableDefault):

    # D-Bus service name, object, and interface to the NetworkManager...
    NETWORK_MANAGER_DBUS_SERVICE_NAME   = "org.freedesktop.NetworkManager"
    NETWORK_MANAGER_DBUS_OBJECT_PATH    = "/org/freedesktop/NetworkManager"
    NETWORK_MANAGER_DBUS_INTERFACE      = "org.freedesktop.NetworkManager"

    # D-Bus NetworkManager state constants...
    NM_STATE_CONNECTED              = 3     # For 0.8 interface
    NM_STATE_CONNECTED_GLOBAL       = 70    # For 0.9 interface

    # Try to get the NetworkManager remote proxy object...
    try:
        
        # Bind to the remote object...
        networkManagerProxy = None
        networkManagerProxy = Gio.DBusProxy.new_for_bus_sync(
            Gio.BusType.SYSTEM,
            Gio.DBusProxyFlags.NONE, 
            None,
            NETWORK_MANAGER_DBUS_SERVICE_NAME,
            NETWORK_MANAGER_DBUS_OBJECT_PATH,
            NETWORK_MANAGER_DBUS_INTERFACE, 
            None)

        # Get the connection state...
        connectionState = networkManagerProxy.state()

    # Something went wrong, but most likely the service isn't available...
    except:
        print(_("Warning: NetworkManager service not found..."))
        return unqueriableDefault

    # Connected...
    if connectionState == NM_STATE_CONNECTED or \
       connectionState == NM_STATE_CONNECTED_GLOBAL:

        # Inform caller...
        return True
    
    # Not connected...
    else:

        # Alert user, if caller requested...
        if alertUser:
            messageDialog = Gtk.MessageDialog(
                parent, Gtk.DialogFlags.MODAL, Gtk.MessageType.ERROR, 
                Gtk.ButtonsType.OK, 
                _("You need a working internet connection in order to do this. "
                "You don't appear to have one right now."))
            messageDialog.run()
            messageDialog.destroy()

        # Inform caller...
        return False

# Open a file or URL as though the user had launched it themselves using their 
#  preferred application...
def launchResource(resourcePath):

    # Winblows...
    if platform.system().lower().find("win") == 0:
        os.startfile(resourcePath)

    # OS X...
    elif platform.system().lower().find("darwin") == 0: 
        subprocess.Popen(["open", format(resourcePath)])
    elif platform.system().lower().find("macosx") == 0: 
        subprocess.Popen(["open", format(resourcePath)])

    # Some other OS...
    else:

        # Try freedesktop approach common on most GNU machines...
        try:
            subprocess.Popen(["xdg-open", format(resourcePath)])
        except OSError:
            print(_("Cannot launch {0}. Platform unknown...").format(resourcePath))

# Open a directory as though the user had launched it themselves using their 
#  preferred shell...
def exploreDirectory(directory):

    # Winblows Explorer...
    if platform.system().lower().find("win") == 0:
        os.startfile(directory, "explore")

    # OS X Finder...
    elif platform.system().lower().find("darwin") == 0: 
        subprocess.Popen(["open", format(directory)])
    elif platform.system().lower().find("macosx") == 0: 
        subprocess.Popen(["open", format(directory)])

    # Some other OS...
    else:

        # Try freedesktop approach common on most GNU machines. This should open
        #  Thunar, Nautilus, or whatever their default shell happens to be...
        try:
            subprocess.Popen(["xdg-open", format(directory)])
        except OSError:
            print(_("Cannot explore {0}. Platform unknown...").format(directory))

